{"nbformat":4,"nbformat_minor":0,"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.1"},"colab":{"name":"Copy of Step_1_7_File Input and Exception Handling.ipynb","provenance":[{"file_id":"1WVlQekjEm2EqLF4QkP2KokYl2TIgkTnl","timestamp":1585774307574},{"file_id":"1yXNNmMK0l0RYHJR1Rl2HJ7TXm88LBDyx","timestamp":1585439476285}]}},"cells":[{"cell_type":"markdown","metadata":{"id":"1UE_BJCHjHtM","colab_type":"text"},"source":["### File Input/Output\n","\n","File is a named location on disk to store related information. When we want to read from or write to a file we need to open it first. When we are done, it needs to be closed, so that resources that are tied with the file are freed.\n","\n","So in Python, a file operation takes place in the following order.\n","1.\tOpen a file\n","2.\tRead or write (perform operation)\n","3.\tClose the file"]},{"cell_type":"code","metadata":{"id":"EkH1AuwEjHtN","colab_type":"code","outputId":"d18bb19a-9bf5-4b57-a287-4aa8c33be66d","colab":{"base_uri":"https://localhost:8080/","height":51},"executionInfo":{"status":"ok","timestamp":1585774747651,"user_tz":-180,"elapsed":7743,"user":{"displayName":"Ibrahim Ghanawi","photoUrl":"","userId":"04665897747668594289"}}},"source":["# Read input from console, input will read numeric values and throw error if a string value entered\n","num = input(\"Enter a number: \")\n","print (type(num))"],"execution_count":1,"outputs":[{"output_type":"stream","text":["Enter a number: 34\n","<class 'str'>\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"zLfeijgfjHtT","colab_type":"text"},"source":["### Writing to a file"]},{"cell_type":"code","metadata":{"id":"YzRNj4qMjHtU","colab_type":"code","outputId":"ca498d0c-259e-4e74-e681-01fd725590c7","colab":{"base_uri":"https://localhost:8080/","height":248},"executionInfo":{"status":"error","timestamp":1585774756222,"user_tz":-180,"elapsed":1038,"user":{"displayName":"Ibrahim Ghanawi","photoUrl":"","userId":"04665897747668594289"}}},"source":["# Set working directory\n","import os\n","\n","# Print the working directory\n","print (\"Current working directory is:\", os.getcwd())\n","\n","# Set working directory\n","os.chdir('C:\\\\Users\\\\Manoh\\\\Documents')\n","\n","# Below code will create a file named vechicles and add the items. \\n is a newline character\n","vechicles = ['scooter\\n', 'bike\\n', 'car\\n']\n","f = open('vechicles.txt', 'w')\n","f.writelines(vechicles)\n","f.close()"],"execution_count":2,"outputs":[{"output_type":"stream","text":["Current working directory is: /content\n"],"name":"stdout"},{"output_type":"error","ename":"FileNotFoundError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-2-3f7faf0ab193>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;31m# Set working directory\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'C:\\\\Users\\\\Manoh\\\\Documents'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m# Below code will create a file named vechicles and add the items. \\n is a newline character\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'C:\\\\Users\\\\Manoh\\\\Documents'"]}]},{"cell_type":"markdown","metadata":{"id":"usGJl2p2jHtW","colab_type":"text"},"source":["### Reading from file"]},{"cell_type":"code","metadata":{"id":"mcHofu4_jHtX","colab_type":"code","outputId":"991fd98e-ec09-4ae7-bbb8-a66ae6b7295b","colab":{"base_uri":"https://localhost:8080/","height":197},"executionInfo":{"status":"error","timestamp":1585774814005,"user_tz":-180,"elapsed":1222,"user":{"displayName":"Ibrahim Ghanawi","photoUrl":"","userId":"04665897747668594289"}}},"source":["f = open('vechicles.txt')\n","print (f.readlines())\n","f.close()"],"execution_count":3,"outputs":[{"output_type":"error","ename":"FileNotFoundError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-3-f767a4856786>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'vechicles.txt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'vechicles.txt'"]}]},{"cell_type":"markdown","metadata":{"id":"z2lZhX2gjHta","colab_type":"text"},"source":["### Exception Handling\n","\n","An exception is an error that happens during the execution of a program that disrupts the normal flow of the program's instructions. When a Python script encounters a situation that it cannot cope with, it raises an exception. Your program should be designed to handle both expected and unexpected errors. "]},{"cell_type":"markdown","metadata":{"id":"TPYGpGqljHta","colab_type":"text"},"source":["#### try and except\n","A critical operation which can raise exception is placed inside the try clause and the code that handles exception is written in except clause"]},{"cell_type":"code","metadata":{"id":"8aEQBhXNjHtb","colab_type":"code","outputId":"3c0217ff-d6df-4816-ca04-84303d1f5b3b","colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"status":"ok","timestamp":1585774831029,"user_tz":-180,"elapsed":1799,"user":{"displayName":"Ibrahim Ghanawi","photoUrl":"","userId":"04665897747668594289"}}},"source":["import sys\n","\n","try:\n","    a = 1\n","    b = 1\n","    print (\"Result of a/b: \", a / b)\n","except (ZeroDivisionError):\n","    print (\"Can't divide by zero\")\n","except (TypeError):\n","    print (\"Wrong data type, division is allowed on numeric data type only\")\n","except:\n","    print (\"Unexpected error occurred\", '\\n', \"Error Type: \", sys.exc_info()[0], '\\n', \"Error Msg: \", sys.exc_info()[1])"],"execution_count":4,"outputs":[{"output_type":"stream","text":["Result of a/b:  1.0\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"FgRIUFs9jHtf","colab_type":"code","outputId":"ef050f38-a964-4a10-fda9-f153c02d942c","colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"status":"ok","timestamp":1585774834714,"user_tz":-180,"elapsed":990,"user":{"displayName":"Ibrahim Ghanawi","photoUrl":"","userId":"04665897747668594289"}}},"source":["try:\n","    a = 1\n","    b = 0\n","    print (a / b)\n","except (ZeroDivisionError):\n","    print (\"Can't divide by zero\")\n","except (TypeError):\n","    print (\"Wrong data type, division is allowed on numeric data type only\")\n","except:\n","    print (\"Unexpected error occurred\", '\\n', \"Error Type: \", sys.exc_info()[0], '\\n', \"Error Msg: \", sys.exc_info()[1])"],"execution_count":5,"outputs":[{"output_type":"stream","text":["Can't divide by zero\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"NwxYjP6DjHti","colab_type":"code","outputId":"0fed38ec-cdeb-490a-98ed-b87298caf5ed","colab":{}},"source":["try:\n","    a = 1\n","    b = 0\n","    print (A / b)\n","except (ZeroDivisionError):\n","    print (\"Can't divide by zero\")\n","except (TypeError):\n","    print (\"Wrong data type, division is allowed on numeric data type only\")\n","except:\n","    print (\"Unexpected error occurred\", '\\n', \"Error Type: \", sys.exc_info()[0], '\\n', \"Error Msg: \", sys.exc_info()[1])"],"execution_count":0,"outputs":[{"output_type":"stream","text":["Unexpected error occurred \n"," Error Type:  <class 'NameError'> \n"," Error Msg:  name 'A' is not defined\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"TzATWdaJjHtl","colab_type":"text"},"source":["#### Finally\n","This is an optional clause which is intended to define clean-up actions that must be executed under all circumstances. "]},{"cell_type":"code","metadata":{"id":"rAaX5rdHjHtl","colab_type":"code","outputId":"4cc712bd-d2c4-442a-ffac-be4a9ded402a","colab":{}},"source":["try:\n","    f = open('C:\\\\Users\\Manoh\\\\Documents\\\\vechicles.txt')\n","    s = f.readline()\n","    print (s)\n","    i = int(s.strip())\n","except IOError as e:\n","    print (\"I/O error({0}): {1}\".format(e.errno, e.strerror))\n","except ValueError:\n","    print (\"Could not convert data to an integer.\")\n","except:\n","    print (\"Unexpected error occurred\", '\\n', \"Error Type: \", sys.exc_info()[0], '\\n', \"Error Msg: \", sys.exc_info()[1])\n","finally:\n","    f.close()\n","    print (\"file has been closed\")"],"execution_count":0,"outputs":[{"output_type":"stream","text":["scooter\n","\n","Could not convert data to an integer.\n","file has been closed\n"],"name":"stdout"}]}]}